In [1]:
import warnings

import numpy as np
import pandas as pd
import useful_rid_code as rid
from openpyxl import load_workbook

warnings.simplefilter("ignore")

# RID 

## Fill-in OSPAR reporting template (parameterised)

This notebook is "parameterised" for use with Papermill. The cell below has the tag `parameters`, which means the entire notebook can be called from `01_recalculate_ospar_1990-2016_main.ipynb`.

In [2]:
# This cell is tagged 'parameters' for use with Papermill
# https://papermill.readthedocs.io/en/latest/index.html
year = 1990
user = ""
pw = ""

In [3]:
# Parameters
user = "jes"
pw = "BeakabusNov21.."
year = 2017


## 1. Get summary data

The notebooks `summary_table_{year}.ipynb` calculate summary values for Table 3. Rather than reading these values from Word, it is easier to repeat the code to summarise the raw data again. This is the data that needs writing to the OSPAR template.

### 1.1. Monitored areas

In [4]:
# Read data
in_csv = f"../../../Results/Loads_CSVs/loads_and_flows_all_sites_{year}.csv"
mon_df = pd.read_csv(in_csv)
del mon_df["new_rid_group"]

# Group by OSPAR region
mon_df1 = mon_df.groupby(["ospar_region", "old_rid_group"]).sum()

# Totals for Norway
mon_df2 = mon_df.groupby("old_rid_group").sum().reset_index()
mon_df2["ospar_region"] = "NORWAY"
mon_df2.set_index(["ospar_region", "old_rid_group"], inplace=True)

# Combine
mon_df = pd.concat([mon_df1, mon_df2], axis=0)

# Cols of interest
cols = [i for i in mon_df.columns if i.split("_")[1] != "Est"]
mon_df = mon_df[cols]
del mon_df["station_id"], mon_df["mean_q_1000m3/day"]

# Convert units
mon_df["Hg_kg"] = mon_df["Hg_kg"] / 1000.0  # kg to tonnes
mon_df["NH4-N_tonnes"] = mon_df["NH4-N_tonnes"] / 1000.0  # tonnes to ktonnes
mon_df["NO3-N_tonnes"] = mon_df["NO3-N_tonnes"] / 1000.0  # tonnes to ktonnes
mon_df["TOTN_tonnes"] = mon_df["TOTN_tonnes"] / 1000.0  # tonnes to ktonnes
mon_df["TOTP_tonnes"] = mon_df["TOTP_tonnes"] / 1000.0  # tonnes to ktonnes
mon_df["PO4-P_tonnes"] = mon_df["PO4-P_tonnes"] / 1000.0  # tonnes to ktonnes
mon_df["SPM_tonnes"] = mon_df["SPM_tonnes"] / 1000.0  # tonnes to ktonnes

# Units are correct, so remove
mon_df.columns = [i.split("_")[0] for i in mon_df.columns]

mon_df.round(0)

Ag    As   Cd    Cr    Cu   Hg  NH4-N  \
ospar_region        old_rid_group                                           
LOFOTEN-BARENTS SEA rid_108        0.0   1.0  0.0   1.0   6.0  0.0    0.0   
                    rid_11         0.0   0.0  0.0   1.0   2.0  0.0    0.0   
                    rid_36         0.0   2.0  0.0   4.0  27.0  0.0    0.0   
NORTH SEA           rid_108        0.0   1.0  0.0   1.0   8.0  0.0    0.0   
                    rid_11         0.0   0.0  0.0   0.0   1.0  0.0    0.0   
                    rid_36         0.0   2.0  0.0   2.0   9.0  0.0    0.0   
NORWEGIAN SEA2      rid_108        0.0   1.0  0.0   1.0   6.0  0.0    0.0   
                    rid_11         0.0   1.0  0.0   1.0  13.0  0.0    0.0   
                    rid_36         0.0   3.0  0.0   9.0  29.0  0.0    0.0   
SKAGERAK            rid_108        0.0   1.0  0.0   1.0   3.0  0.0    0.0   
                    rid_11         0.0   8.0  1.0   9.0  50.0  0.0    1.0   
                    rid_36         0.0   2.0  0.0   0.0   6.0  0.0    0.0   
NORWAY              rid_108        0.0   4.0  0.0   4.0  22.0  0.0    0.0   
                    rid_11         0.0  10.0  1.0  11.0  66.0  0.0    1.0   
                    rid_36         0.0   8.0  1.0  16.0  70.0  0.0    1.0   

                                   NO3-N    Ni  PO4-P    Pb    SPM      SiO2  \
ospar_region        old_rid_group                                              
LOFOTEN-BARENTS SEA rid_108          0.0   5.0    0.0   0.0   10.0   22962.0   
                    rid_11           0.0   1.0    0.0   0.0    7.0   16367.0   
                    rid_36           0.0  56.0    0.0   1.0   51.0   80175.0   
NORTH SEA           rid_108          3.0   5.0    0.0   3.0   19.0   21243.0   
                    rid_11           0.0   1.0    0.0   0.0    5.0    3989.0   
                    rid_36           3.0   3.0    0.0   3.0   41.0   29231.0   
NORWEGIAN SEA2      rid_108          1.0   4.0    0.0   1.0   23.0   19356.0   
                    rid_11           0.0   3.0    0.0   0.0    8.0   10405.0   
                    rid_36           3.0  22.0    0.0   2.0   77.0   63813.0   
SKAGERAK            rid_108          1.0   2.0    0.0   1.0   13.0    8729.0   
                    rid_11          14.0  28.0    0.0  10.0  268.0  164223.0   
                    rid_36           1.0   3.0    0.0   3.0   32.0   21098.0   
NORWAY              rid_108          5.0  17.0    0.0   5.0   65.0   72290.0   
                    rid_11          15.0  33.0    0.0  10.0  287.0  194984.0   
                    rid_36           7.0  84.0    0.0   9.0  201.0  194316.0   

                                        TOC  TOTN  TOTP     Zn  
ospar_region        old_rid_group                               
LOFOTEN-BARENTS SEA rid_108         26049.0   1.0   0.0    6.0  
                    rid_11          12030.0   1.0   0.0    1.0  
                    rid_36          64587.0   4.0   0.0   26.0  
NORTH SEA           rid_108         22860.0   4.0   0.0   23.0  
                    rid_11           5335.0   1.0   0.0    4.0  
                    rid_36          44727.0   6.0   0.0   41.0  
NORWEGIAN SEA2      rid_108         24186.0   2.0   0.0    9.0  
                    rid_11          11981.0   1.0   0.0   21.0  
                    rid_36          71282.0   6.0   0.0   69.0  
SKAGERAK            rid_108         15250.0   2.0   0.0   19.0  
                    rid_11         209470.0  23.0   1.0  124.0  
                    rid_36          59426.0   3.0   0.0   32.0  
NORWAY              rid_108         88346.0   9.0   0.0   57.0  
                    rid_11         238816.0  25.0   1.0  150.0  
                    rid_36         240022.0  19.0   1.0  168.0

### 1.2. Unmonitored areas

In [5]:
# Read data
in_csv = f"../../../Results/Unmon_loads/teotil2_ospar_unmonitored_loads_{year}.csv"
umon_df = pd.read_csv(in_csv, index_col=0)

# Rename cols
umon_df.columns = [i.replace("RENSEANLEGG", "sew") for i in umon_df.columns]
umon_df.columns = [i.replace("INDUSTRI", "ind") for i in umon_df.columns]
umon_df.columns = [i.replace("_tonn", "") for i in umon_df.columns]
umon_df.columns = [i.replace("AQUAKULTUR", "fish") for i in umon_df.columns]

# Add missing columns if necessary
cols = ["sew_Hg", "sew_S.P.M.", "ind_Hg"]
for col in cols:
    if col not in umon_df.columns:
        umon_df[col] = 0

# Convert Hg to kgs
umon_df["sew_Hg"] = umon_df["sew_Hg"] * 1000
umon_df["ind_Hg"] = umon_df["ind_Hg"] * 1000

umon_df.round(0)

flow    sew_n   sew_p   ind_n  ind_p   fish_n  \
ospar_region                                                             
NORWAY               417168.0  14015.0  1079.0  2459.0  290.0  59962.0   
LOFOTEN-BARENTS SEA  121985.0   1218.0   165.0    81.0    6.0  15814.0   
NORTH SEA            129896.0   3751.0   434.0   517.0   77.0  19137.0   
NORWEGIAN SEA2       156445.0   2748.0   353.0   911.0  158.0  24982.0   
SKAGERAK               8843.0   6298.0   128.0   951.0   50.0     29.0   

                      fish_p   diff_n  diff_p  sew_po4  ...  sew_S.P.M.  \
ospar_region                                            ...               
NORWAY               10350.0  34965.0   712.0    648.0  ...     13676.0   
LOFOTEN-BARENTS SEA   2731.0   5630.0   112.0     99.0  ...      2382.0   
NORTH SEA             3299.0  13430.0   225.0    260.0  ...      5429.0   
NORWEGIAN SEA2        4314.0  12981.0   288.0    212.0  ...      3546.0   
SKAGERAK                 5.0   2925.0    87.0     77.0  ...      2319.0   

                     sew_As  sew_Pb  sew_Cd  sew_Cu  sew_Zn  sew_Ni  sew_Cr  \
ospar_region                                                                  
NORWAY                  0.0     0.0     0.0     5.0    20.0     3.0     1.0   
LOFOTEN-BARENTS SEA     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
NORTH SEA               0.0     0.0     0.0     2.0     9.0     1.0     1.0   
NORWEGIAN SEA2          0.0     0.0     0.0     1.0     1.0     0.0     0.0   
SKAGERAK                0.0     0.0     0.0     2.0     9.0     1.0     0.0   

                     sew_Hg  fish_Cu  
ospar_region                          
NORWAY                 10.0   1217.0  
LOFOTEN-BARENTS SEA     0.0    321.0  
NORTH SEA               4.0    387.0  
NORWEGIAN SEA2          3.0    509.0  
SKAGERAK                2.0      1.0  

[5 rows x 41 columns]

### 1.3. Loads for 11 main rivers

In [6]:
# Read data
in_csv = f"../../../Results/Loads_CSVs/loads_and_flows_all_sites_{year}.csv"
rid11_df = pd.read_csv(in_csv, index_col=0)
del rid11_df["new_rid_group"]

# Get data for RID11
rid11_df = rid11_df.query('old_rid_group == "rid_11"')

# Tidy
del rid11_df["station_code"], rid11_df["station_name"]
del rid11_df["old_rid_group"], rid11_df["ospar_region"]
del rid11_df["mean_q_1000m3/day"]

cols = [i for i in rid11_df.columns if i.split("_")[1] != "Est"]
rid11_df = rid11_df[cols]

# Convert units
rid11_df["Hg_kg"] = rid11_df["Hg_kg"] / 1000.0  # kg to tonnes
rid11_df["NH4-N_tonnes"] = rid11_df["NH4-N_tonnes"] / 1000.0  # tonnes to ktonnes
rid11_df["NO3-N_tonnes"] = rid11_df["NO3-N_tonnes"] / 1000.0  # tonnes to ktonnes
rid11_df["TOTN_tonnes"] = rid11_df["TOTN_tonnes"] / 1000.0  # tonnes to ktonnes
rid11_df["TOTP_tonnes"] = rid11_df["TOTP_tonnes"] / 1000.0  # tonnes to ktonnes
rid11_df["PO4-P_tonnes"] = rid11_df["PO4-P_tonnes"] / 1000.0  # tonnes to ktonnes
rid11_df["SPM_tonnes"] = rid11_df["SPM_tonnes"] / 1000.0  # tonnes to ktonnes

# Tidy cols
rid11_df.columns = [i.split("_")[0] for i in rid11_df.columns]

rid11_df.head()

Ag        As        Cd        Cr        Cu        Hg  \
station_id                                                               
29612       0.000000  1.885934  0.091354  1.704438  8.229109  0.012560   
29779       0.000000  0.413012  0.000000  0.682191  1.719991  0.003129   
29821       0.000000  0.207549  0.013841  0.162546  1.044845  0.000000   
29782       0.000000  0.359566  0.009270  0.381896  1.064020  0.000000   
36225       0.000975  0.026755  0.005583  0.095556  0.426425  0.000067   

               NH4-N     NO3-N        Ni     PO4-P        Pb        SPM  \
station_id                                                                
29612       0.103427  2.618654  5.265423  0.026224  1.444168  20.771381   
29779       0.012023  0.194927  0.870596  0.015467  0.044949   6.845721   
29821       0.015083  0.307519  0.815021  0.003429  0.158681   1.942288   
29782       0.003611  0.094011  0.982699  0.003345  0.191518   4.889757   
36225       0.003227  0.036341  0.113191  0.002043  0.161677   0.433375   

                    SiO2           TOC      TOTN      TOTP         Zn  
station_id                                                             
29612       31359.804331  41294.356441  4.256075  0.067885  24.325649  
29779       16367.010804  12029.950310  0.704780  0.026165   0.576850  
29821        3063.193676   3805.132158  0.500895  0.009256   2.829575  
29782        4305.804503   5793.362515  0.327468  0.011089   1.047347  
36225         290.341819    200.175307  0.060726  0.002603   1.853462

## 2. Fill-in template

The template is usually sent each year by Csilla at NIBIO. However, it doesn't seem to change, so old versions can also be used e.g. here:

    ../../../Results/OSPAR/Blank_Template

In [7]:
# Copy of template to update
temp_path = rid.copy_ospar_template(year)

In [8]:
def update_spreadsheet_point_sources(xlsx, sheet, pars, src, df):
    """Update the OSPAR template for point source data.

    Args:
        xslx:  Str. Path to Excel template
        sheet: Str. Sheet name to update
        pars:  List. Parameter names in template to fill-in
        src:   Str. Type of input ('sew', 'ind', 'fish')
        df:    Dataframe. Values to fill-in

    Returns:
        None. The template is updated and saved.
    """
    import pandas as pd
    from openpyxl import load_workbook

    # Map Excel headings to df cols
    par_dict = {
        "SPM": "S.P.M.",
        "TOC": "TOC",
        "PO4-P": "po4",
        "P-Total": "p",
        "NO3-N": "no3",
        "NH4-N": "nh4",
        "N-Total": "n",
        "As": "As",
        "Pb": "Pb",
        "Cd": "Cd",
        "Cu": "Cu",
        "Zn": "Zn",
        "Ni": "Ni",
        "Total Cr": "Cr",
        "Hg": "Hg",
    }

    # Map template names to df names
    names_dict = {
        "Norwegian Sea (NO)": "NORWEGIAN SEA2",
        "Barents Sea (NO)": "LOFOTEN-BARENTS SEA",
        "Skagerrak (NO)": "SKAGERAK",
        "North Sea (NO)": "NORTH SEA",
        "Norway Total": "NORWAY",
    }

    # Open new file and get sheet
    wb = load_workbook(filename=xlsx)
    ws = wb[sheet]

    # Get row numbers
    row_dict = {}
    for item in ws["B12" : "B%s" % ws.max_row]:
        # Get cell properties
        cell = item[0]
        name = cell.value
        row = cell.row
        row_dict[name] = row

    # Get col numbers
    col_dict = {}
    for cell in ws["E9":"AK9"][0]:
        # Get cell properties
        par = cell.value
        col = cell.column
        col_dict[par] = col

    # Update spreadsheet
    for reg in names_dict.keys():
        for par in pars:
            # Get value from df
            try:
                val = df.loc[names_dict[reg], "%s_%s" % (src, par_dict[par])]
            except KeyError:
                val = 0

            # Get cell co-ords
            row = row_dict[reg] + 2
            col = col_dict[par]

            # Write value
            ws.cell(row=row, column=col).value = val

    # Save
    wb.save(xlsx)

### 2.1. Sheet 5a: Sewage effluents

Note these are the values from Table 3 for **unmonitored areas** and not the total sewage inputs for each of the OSPAR areas. This is the same as what Tore reported previously.

In [9]:
# Get Sewage data
cols = [i for i in umon_df.columns if i.split("_")[0] == "sew"]
sew_df = umon_df[cols].copy()

# Convert units
sew_df["sew_Hg"] = sew_df["sew_Hg"] / 1000.0  # kg to tonnes
sew_df["sew_nh4"] = sew_df["sew_nh4"] / 1000.0  # tonnes to ktonnes
sew_df["sew_no3"] = sew_df["sew_no3"] / 1000.0  # tonnes to ktonnes
sew_df["sew_n"] = sew_df["sew_n"] / 1000.0  # tonnes to ktonnes
sew_df["sew_po4"] = sew_df["sew_po4"] / 1000.0  # tonnes to ktonnes
sew_df["sew_p"] = sew_df["sew_p"] / 1000.0  # tonnes to ktonnes
sew_df["sew_S.P.M."] = sew_df["sew_S.P.M."] / 1000.0  # tonnes to ktonnes

sew_df

sew_n     sew_p   sew_po4   sew_no3    sew_nh4  \
ospar_region                                                              
NORWAY               14.015213  1.079347  0.647608  0.700761  10.511409   
LOFOTEN-BARENTS SEA   1.218243  0.164823  0.098894  0.060912   0.913682   
NORTH SEA             3.751038  0.433758  0.260255  0.187552   2.813279   
NORWEGIAN SEA2        2.747626  0.352555  0.211533  0.137381   2.060719   
SKAGERAK              6.298306  0.128210  0.076926  0.314915   4.723729   

                     sew_S.P.M.    sew_As    sew_Pb    sew_Cd    sew_Cu  \
ospar_region                                                              
NORWAY                13.675806  0.382209  0.402408  0.024947  5.422717   
LOFOTEN-BARENTS SEA    2.381861  0.003655  0.005070  0.000226  0.077380   
NORTH SEA              5.429303  0.174091  0.214789  0.012058  2.258631   
NORWEGIAN SEA2         3.545793  0.041571  0.033790  0.001239  0.746265   
SKAGERAK               2.318848  0.162892  0.148759  0.011424  2.340441   

                        sew_Zn    sew_Ni    sew_Cr    sew_Hg  
ospar_region                                                  
NORWAY               19.828626  2.672803  1.249492  0.009769  
LOFOTEN-BARENTS SEA   0.214856  0.019237  0.010721  0.000126  
NORTH SEA             9.158609  1.165917  0.670089  0.004399  
NORWEGIAN SEA2        1.440710  0.130597  0.348258  0.002907  
SKAGERAK              9.014451  1.357052  0.220424  0.002337

In [10]:
# Update sheet 5a
pars = [
    "Ni",
    "Pb",
    "NH4-N",
    "Total Cr",
    "NO3-N",
    "Zn",
    "As",
    "Cd",
    "P-Total",
    "SPM",
    "PO4-P",
    "N-Total",
    "Hg",
    "Cu",
]

update_spreadsheet_point_sources(temp_path, "5a", pars, "sew", sew_df)

### 2.2. Sheet 5b: Industrial effluents

In [11]:
# Get industrial data
cols = [i for i in umon_df.columns if i.split("_")[0] == "ind"]
ind_df = umon_df[cols].copy()

# Convert units
ind_df["ind_Hg"] = ind_df["ind_Hg"] / 1000.0  # kg to tonnes
ind_df["ind_nh4"] = ind_df["ind_nh4"] / 1000.0  # tonnes to ktonnes
ind_df["ind_no3"] = ind_df["ind_no3"] / 1000.0  # tonnes to ktonnes
ind_df["ind_n"] = ind_df["ind_n"] / 1000.0  # tonnes to ktonnes
ind_df["ind_po4"] = ind_df["ind_po4"] / 1000.0  # tonnes to ktonnes
ind_df["ind_p"] = ind_df["ind_p"] / 1000.0  # tonnes to ktonnes
ind_df["ind_S.P.M."] = ind_df["ind_S.P.M."] / 1000.0  # tonnes to ktonnes

ind_df

ind_n     ind_p   ind_po4   ind_no3   ind_nh4  \
ospar_region                                                            
NORWAY               2.458854  0.290106  0.174063  0.122943  1.844141   
LOFOTEN-BARENTS SEA  0.081010  0.005899  0.003539  0.004051  0.060758   
NORTH SEA            0.516798  0.076800  0.046080  0.025840  0.387598   
NORWEGIAN SEA2       0.910539  0.157878  0.094727  0.045527  0.682904   
SKAGERAK             0.950508  0.049529  0.029718  0.047525  0.712881   

                     ind_S.P.M.    ind_TOC    ind_As    ind_Pb    ind_Cd  \
ospar_region                                                               
NORWAY               453.779847  1001.7977  1.940754  1.095565  0.080103   
LOFOTEN-BARENTS SEA  232.754320   370.7891  0.000141  0.000029  0.000008   
NORTH SEA             22.954929   463.7137  1.519400  0.726108  0.054004   
NORWEGIAN SEA2       196.629785    96.8410  0.005302  0.105784  0.011335   
SKAGERAK               1.440813    70.4539  0.415911  0.263644  0.014757   

                       ind_Cu     ind_Zn    ind_Ni    ind_Cr    ind_Hg  
ospar_region                                                            
NORWAY               5.109014  15.674884  6.450159  1.473708  0.009469  
LOFOTEN-BARENTS SEA  0.000572   0.001133  0.000288  0.000069  0.000109  
NORTH SEA            0.620646   4.456207  4.342587  0.715340  0.002716  
NORWEGIAN SEA2       0.251153   4.330596  0.113109  0.061616  0.000471  
SKAGERAK             4.236643   6.886947  1.994176  0.696682  0.006174

In [12]:
# Update sheet 5b
pars = [
    "Ni",
    "Pb",
    "NH4-N",
    "Total Cr",
    "NO3-N",
    "Zn",
    "As",
    "Cd",
    "P-Total",
    "SPM",
    "PO4-P",
    "N-Total",
    "Hg",
    "Cu",
    "TOC",
]

update_spreadsheet_point_sources(temp_path, "5b", pars, "ind", ind_df)

### 2.3. Sheet 5c: Aquaculture discharges

In [13]:
# Get fish data
cols = [i for i in umon_df.columns if i.split("_")[0] == "fish"]
fish_df = umon_df[cols].copy()

# Convert units
fish_df["fish_nh4"] = fish_df["fish_nh4"] / 1000.0  # tonnes to ktonnes
fish_df["fish_no3"] = fish_df["fish_no3"] / 1000.0  # tonnes to ktonnes
fish_df["fish_n"] = fish_df["fish_n"] / 1000.0  # tonnes to ktonnes
fish_df["fish_po4"] = fish_df["fish_po4"] / 1000.0  # tonnes to ktonnes
fish_df["fish_p"] = fish_df["fish_p"] / 1000.0  # tonnes to ktonnes

fish_df

fish_n     fish_p  fish_po4  fish_no3   fish_nh4  \
ospar_region                                                               
NORWAY               59.962278  10.349808  7.141367  6.595851  47.969823   
LOFOTEN-BARENTS SEA  15.813914   2.730805  1.884255  1.739531  12.651131   
NORTH SEA            19.136870   3.299401  2.276586  2.105056  15.309496   
NORWEGIAN SEA2       24.982000   4.314385  2.976926  2.748020  19.985600   
SKAGERAK              0.029494   0.005217  0.003600  0.003244   0.023595   

                         fish_Cu  
ospar_region                      
NORWAY               1217.200000  
LOFOTEN-BARENTS SEA   320.509412  
NORTH SEA             387.244422  
NORWEGIAN SEA2        508.833824  
SKAGERAK                0.612342

In [14]:
# Update sheet 5c
pars = ["NH4-N", "NO3-N", "P-Total", "PO4-P", "N-Total", "Cu"]

update_spreadsheet_point_sources(temp_path, "5c", pars, "fish", fish_df)

### 2.4. Sheet 5d: Other discharges

This sheet is left blank

### 2.5. Sheet 5e: Total direct discharges

The sum of sewage, industrial and fish-farm discharges.

In [15]:
# Combine sew, ind and fish, then aggregate
for df in [sew_df, ind_df, fish_df]:
    df.reset_index(inplace=True)
    df.columns = [i.split("_")[1] for i in df.columns]

td_df = pd.concat([sew_df, ind_df, fish_df], axis=0, sort=True)
td_df = td_df.groupby("region").sum()

td_df.columns = ["td_" + i for i in td_df.columns]

td_df

td_As     td_Cd     td_Cr        td_Cu     td_Hg  \
region                                                                     
LOFOTEN-BARENTS SEA  0.003796  0.000234  0.010790   320.587364  0.000235   
NORTH SEA            1.693491  0.066062  1.385429   390.123699  0.007115   
NORWAY               2.322963  0.105050  2.723200  1227.731731  0.019238   
NORWEGIAN SEA2       0.046873  0.012574  0.409874   509.831242  0.003378   
SKAGERAK             0.578803  0.026181  0.917106     7.189426  0.008511   

                        td_Ni     td_Pb   td_S.P.M.     td_TOC      td_Zn  \
region                                                                      
LOFOTEN-BARENTS SEA  0.019525  0.005099  235.136180   370.7891   0.215989   
NORTH SEA            5.508504  0.940897   28.384233   463.7137  13.614816   
NORWAY               9.122962  1.497973  467.455652  1001.7977  35.503510   
NORWEGIAN SEA2       0.243706  0.139574  200.175578    96.8410   5.771306   
SKAGERAK             3.351228  0.412403    3.759661    70.4539  15.901398   

                          td_n     td_nh4    td_no3       td_p    td_po4  
region                                                                    
LOFOTEN-BARENTS SEA  17.113167  13.625571  1.804493   2.901527  1.986689  
NORTH SEA            23.404706  18.510373  2.318448   3.809959  2.582921  
NORWAY               76.436345  60.325373  7.419554  11.719260  7.963039  
NORWEGIAN SEA2       28.640164  22.729223  2.930928   4.824818  3.283185  
SKAGERAK              7.278308   5.460205  0.365685   0.182957  0.110244

In [16]:
# Update sheet 5e
pars = [
    "Ni",
    "Pb",
    "NH4-N",
    "Total Cr",
    "NO3-N",
    "Zn",
    "As",
    "Cd",
    "P-Total",
    "SPM",
    "PO4-P",
    "N-Total",
    "Hg",
    "Cu",
    "TOC",
]

update_spreadsheet_point_sources(temp_path, "5e", pars, "td", td_df)

### 2.6. Sheet 6a: Monitored rivers

**Does "Inner Oslofjord" in the template correspond to "Alna"?** If so, I can fill-in one additional row in this table.

In [17]:
mon_df.reset_index(inplace=True)
tot_df = mon_df.groupby("ospar_region").sum()
trib_df = mon_df[mon_df["old_rid_group"] != "rid_11"].groupby("ospar_region").sum()

In [18]:
def update_spreadsheet_monitored_rivers(xlsx, sheet, pars, df_dict):
    """Update the OSPAR template for monitored rivers.

    Args:
        xslx:    Str. Path to Excel template
        sheet:   Str. Sheet name to update
        pars:    List. Parameter names in template to fill-in
        df_dict: Dict. {'tot':tot_df, 'trib':trib_df, 'main':rid11_df}
                 Values to fill-in

    Returns:
        None. The template is updated and saved.
    """
    import pandas as pd
    from openpyxl import load_workbook

    # Map Excel headings to df cols
    par_dict = {
        "SPM": "SPM",
        "TOC": "TOC",
        "PO4-P": "PO4-P",
        "P-Total": "TOTP",
        "NO3-N": "NO3-N",
        "NH4-N": "NH4-N",
        "N-Total": "TOTN",
        "As": "As",
        "Pb": "Pb",
        "Cd": "Cd",
        "Cu": "Cu",
        "Zn": "Zn",
        "Ni": "Ni",
        "Total Cr": "Cr",
        "Hg": "Hg",
    }

    # Map template names to df names and rows
    names_dict = {
        "Norwegian Sea (NO)": ("NORWEGIAN SEA2", "tot"),
        "Barents Sea (NO)": ("LOFOTEN-BARENTS SEA", "tot"),
        "Skagerrak (NO)": ("SKAGERAK", "tot"),
        "North Sea (NO)": ("NORTH SEA", "tot"),
        "Norway Total": ("NORWAY", "tot"),
        "Tributary Rivers - Norwegian Sea": ("NORWEGIAN SEA2", "trib"),
        "Tributary Rivers - Barents Sea": ("LOFOTEN-BARENTS SEA", "trib"),
        "Tributary Rivers - Skagerak": ("SKAGERAK", "trib"),
        "Tributary Rivers - North Sea": ("NORTH SEA", "trib"),
        "Orkla": (29778, "main"),
        "Vefsna": (29782, "main"),
        "Alta": (29779, "main"),
        "Glomma": (29617, "main"),
        "Drammenselva": (29612, "main"),
        u"Numedalslågen": (29615, "main"),
        "Skienselva": (29613, "main"),
        "Otra": (29614, "main"),
        "Orreelva": (29783, "main"),
        "Vosso": (29821, "main"),
    }

    # Open new file and get sheet
    wb = load_workbook(filename=xlsx)
    ws = wb[sheet]

    # Get row numbers
    row_dict = {}
    for item in ws["B12" : "B%s" % ws.max_row]:
        # Get cell properties
        cell = item[0]
        name = cell.value
        row = cell.row
        row_dict[name] = row

    # Get col numbers
    col_dict = {}
    for cell in ws["E9":"AK9"][0]:
        # Get cell properties
        par = cell.value
        col = cell.column
        col_dict[par] = col

    # Update spreadsheet
    for reg in names_dict.keys():
        df_idx, df_name = names_dict[reg]

        # Get df
        df = df_dict[df_name]

        for par in pars:
            # Get value from df
            val = df.loc[df_idx, par_dict[par]]

            # Get cell co-ords
            row = row_dict[reg] + 2
            col = col_dict[par]

            # Write value
            ws.cell(row=row, column=col).value = val

    # Save
    wb.save(xlsx)

In [19]:
# Update sheet 6a
pars = [
    "Ni",
    "Pb",
    "NH4-N",
    "Total Cr",
    "NO3-N",
    "Zn",
    "As",
    "Cd",
    "P-Total",
    "SPM",
    "PO4-P",
    "N-Total",
    "Hg",
    "Cu",
    "TOC",
]
df_dict = {"tot": tot_df, "main": rid11_df, "trib": trib_df}

update_spreadsheet_monitored_rivers(temp_path, "6a", pars, df_dict)

### 2.7. Sheet 6b: Unmonitored areas

In [20]:
# Get diff data
cols = [i for i in umon_df.columns if i.split("_")[0] == "diff"]
diff_df = umon_df[cols].copy()

# Convert units
diff_df["diff_nh4"] = diff_df["diff_nh4"] / 1000.0  # tonnes to ktonnes
diff_df["diff_no3"] = diff_df["diff_no3"] / 1000.0  # tonnes to ktonnes
diff_df["diff_n"] = diff_df["diff_n"] / 1000.0  # tonnes to ktonnes
diff_df["diff_po4"] = diff_df["diff_po4"] / 1000.0  # tonnes to ktonnes
diff_df["diff_p"] = diff_df["diff_p"] / 1000.0  # tonnes to ktonnes

diff_df

diff_n    diff_p  diff_po4   diff_no3  diff_nh4
ospar_region                                                           
NORWAY               34.965240  0.712162  0.175192  21.853275  1.923088
LOFOTEN-BARENTS SEA   5.629640  0.112171  0.027594   3.518525  0.309630
NORTH SEA            13.429547  0.224938  0.055335   8.393467  0.738625
NORWEGIAN SEA2       12.981052  0.287554  0.070738   8.113158  0.713958
SKAGERAK              2.925000  0.087500  0.021525   1.828125  0.160875

In [21]:
# Update sheet 6b
pars = ["NH4-N", "NO3-N", "P-Total", "PO4-P", "N-Total"]

update_spreadsheet_point_sources(temp_path, "6b", pars, "diff", diff_df)

### 2.8. Sheet 6c: Total inputs

**Note:** See e-mail from Csilla received 07/11/2017 at 13.43. This table should **not** include "point" discharges (`td_df`) - it's just the sum of tables 6a and 6b.

In [22]:
# Rename cols in diff_df
col_map = {
    "diff_n": "TOTN",
    "diff_p": "TOTP",
    "diff_po4": "PO4-P",
    "diff_no3": "NO3-N",
    "diff_nh4": "NH4-N",
}
diff_df.columns = [col_map[i] for i in diff_df.columns]

# Add to total_df
for col in diff_df.columns:
    tot_df[col] = tot_df[col] + diff_df[col]

# Update sheet 6c
pars = [
    "Ni",
    "Pb",
    "NH4-N",
    "Total Cr",
    "NO3-N",
    "Zn",
    "As",
    "Cd",
    "P-Total",
    "SPM",
    "PO4-P",
    "N-Total",
    "Hg",
    "Cu",
    "TOC",
]

df_dict = {"tot": tot_df, "main": rid11_df, "trib": trib_df}

update_spreadsheet_monitored_rivers(temp_path, "6c", pars, df_dict)

The following two cells are not longer used as I originally misunderstood what table 6c represents.

In [23]:
## Standardise col names
# diff_df.columns = [i.split('_')[1] for i in diff_df.columns]
# td_df.columns = [i.split('_')[1] for i in td_df.columns]
# td_df.index.name = 'ospar_region'
#
# col_dict = {'SPM':'S.P.M.',
#            'TOTN':'n',
#            'NH4-N':'nh4',
#            'NO3-N':'no3',
#            'TOTP':'p',
#            'PO4-P':'po4'}
# for col in col_dict.keys():
#    new_col = col_dict[col]
#    tot_df[new_col] = tot_df[col]
#    del tot_df[col]
#
## Reset index
# diff_df.reset_index(inplace=True)
# tot_df.reset_index(inplace=True)
# td_df.reset_index(inplace=True)
#
## Concat and aggregate
##tot_df = pd.concat([diff_df, td_df, tot_df], axis=0).groupby('ospar_region').sum() # See comment above
# tot_df = pd.concat([diff_df, tot_df], axis=0).groupby('ospar_region').sum()
#
## Rename cols
# tot_df.columns = ['tot_'+i for i in tot_df.columns]
#
# tot_df

In [24]:
## Update sheet 6c
# pars = ['Ni', 'Pb', 'NH4-N', 'Total Cr', 'NO3-N',
#        'Zn', 'As', 'Cd', 'P-Total', 'SPM', 'PO4-P',
#        'N-Total', 'Hg', 'Cu', 'TOC']
#
# update_spreadsheet_point_sources(temp_path, '6c', pars, 'tot', tot_df)

### 2.9. Sheet 7: Concentrations

In [25]:
# Read data
in_csv = f"../../../Results/Loads_CSVs/concs_and_flows_rid_11_{year}.csv"
conc_df = pd.read_csv(in_csv, index_col=0, encoding="utf-8")

cols = [
    "Hg_ng/l",
    "NH4-N_µg/l N",
    "NO3-N_µg/l N",
    "TOTN_µg/l N",
    "TOTP_µg/l P",
    "PO4-P_µg/l P",
    "TOC_mg C/l",
]
for col in cols:
    if col not in conc_df.columns:
        par = col.split("_")[0]
        conc_df[col] = np.nan
        conc_df[f"{par}_flag"] = np.nan

# Convert units
conc_df["Hg_ng/l"] = conc_df["Hg_ng/l"] / 1000  # ng to ug
conc_df["NH4-N_µg/l N"] = conc_df["NH4-N_µg/l N"] / 1000  # ug to mg
conc_df["NO3-N_µg/l N"] = conc_df["NO3-N_µg/l N"] / 1000  # ug to mg
conc_df["TOTN_µg/l N"] = conc_df["TOTN_µg/l N"] / 1000  # ug to mg
conc_df["TOTP_µg/l P"] = conc_df["TOTP_µg/l P"] / 1000  # ug to mg
conc_df["PO4-P_µg/l P"] = conc_df["PO4-P_µg/l P"] / 1000  # ug to mg
conc_df["TOC_mg C/l"] = conc_df["TOC_mg C/l"] * 1000  # mg to ug

# Get flags
cols = [i for i in conc_df.columns if i.split("_")[1] == "flag"]
lod_df = conc_df[cols]
lod_df.columns = [i.split("_")[0] for i in lod_df.columns]

# Get vals
cols = [
    i
    for i in conc_df.columns
    if ((i.split("_")[0] in lod_df.columns) and (i.split("_")[1] != "flag"))
]
conc_df = conc_df[cols]
conc_df.columns = [i.split("_")[0] for i in conc_df.columns]

# Rename
col_dict = {
    "SPM": "S.P.M.",
    "TOTN": "n",
    "NH4-N": "nh4",
    "NO3-N": "no3",
    "TOTP": "p",
    "PO4-P": "po4",
}
for col in col_dict.keys():
    new_col = col_dict[col]
    lod_df[new_col] = lod_df[col]
    conc_df[new_col] = conc_df[col]
    del conc_df[col], lod_df[col]

# Map Excel headings to df cols
par_dict = {
    "SPM": "S.P.M.",
    "TOC": "TOC",
    "PO4-P": "po4",
    "P-Total": "p",
    "NO3-N": "no3",
    "NH4-N": "nh4",
    "N-Total": "n",
    "As": "As",
    "Pb": "Pb",
    "Cd": "Cd",
    "Cu": "Cu",
    "Zn": "Zn",
    "Ni": "Ni",
    "Total Cr": "Cr",
    "Hg": "Hg",
}

# Map names to stns
names_dict = {
    "Orkla": 29778,
    "Vefsna": 29782,
    "Alta": 29779,
    "Glomma": 29617,
    "Drammenselva": 29612,
    "Numedalslågen": 29615,
    "Skienselva": 29613,
    "Otra": 29614,
    "Orreelva": 29783,
    "Vosso": 29821,
}

# Open new file and get sheet
wb = load_workbook(filename=temp_path)
ws = wb["7"]

# Get row numbers
row_dict = {}
for item in ws["B12" : "B%s" % ws.max_row]:
    # Get cell properties
    cell = item[0]
    name = cell.value
    row = cell.row
    row_dict[name] = row

# Get col numbers
col_dict = {}
for cell in ws["E9":"AK9"][0]:
    # Get cell properties
    par = cell.value
    col = cell.column
    col_dict[par] = col

pars = [
    "Ni",
    "Pb",
    "NH4-N",
    "Total Cr",
    "NO3-N",
    "Zn",
    "As",
    "Cd",
    "P-Total",
    "SPM",
    "PO4-P",
    "N-Total",
    "Hg",
    "Cu",
    "TOC",
]

# Update spreadsheet
for reg in names_dict.keys():
    for par in pars:
        if par_dict[par] not in conc_df.columns:
            conc_df[par_dict[par]] = np.nan
            lod_df[par_dict[par]] = np.nan

        # Get values from df
        # 1a. Lower average
        vals = conc_df.loc[names_dict[reg]].copy()[[par_dict[par]]].values
        lods = lod_df.loc[names_dict[reg]].copy()[[par_dict[par]]].fillna("0").values
        vals[(lods == "<")] = 0
        val = np.nanmean(vals)

        row = row_dict[reg]
        col = col_dict[par]
        ws.cell(row=row, column=col).value = val

        # 1b. Upper average
        val = conc_df.loc[names_dict[reg], par_dict[par]].mean()
        row = row_dict[reg]
        col = col_dict[par]
        ws.cell(row=row, column=col).value = val

        # 2. Min
        val = conc_df.loc[names_dict[reg], par_dict[par]].min()
        row = row_dict[reg] + 2
        col = col_dict[par]
        ws.cell(row=row, column=col).value = val

        # 3. Max
        val = conc_df.loc[names_dict[reg], par_dict[par]].max()
        row = row_dict[reg] + 3
        col = col_dict[par]
        ws.cell(row=row, column=col).value = val

        # 4. N
        n_samp = len(conc_df.loc[names_dict[reg]].copy()[[par_dict[par]]].dropna())
        if n_samp == 0:
            n_samp = np.nan
        row = row_dict[reg] + 5
        col = col_dict[par]
        ws.cell(row=row, column=col).value = n_samp

        # 5. N LOD
        n_lod = (~pd.isnull(lod_df.loc[names_dict[reg]].copy()[[par_dict[par]]])).sum()
        try:
            pct_lod = 100 * float(n_lod) / float(n_samp)
            if pct_lod < 30:
                val = "Yes"
            else:
                val = "No"
        except ZeroDivisionError:
            val = np.nan

        row = row_dict[reg] + 4
        col = col_dict[par]
        ws.cell(row=row, column=col).value = val

        # 6. Std
        val = conc_df.loc[names_dict[reg], par_dict[par]].std()
        row = row_dict[reg] + 7
        col = col_dict[par]
        ws.cell(row=row, column=col).value = val

# Save
wb.save(temp_path)

### 2.10. Sheet 9: Discharge

The notebook [here](http://nbviewer.jupyter.org/github/JamesSample/rid/blob/master/notebooks/recalculate_ospar_flows.ipynb) handles the OSPAR flow data. Run this, and then copy the results over to Sheet 9. 